src/pages/LoginPage.js

```diff
@@ -1,16 +1,102 @@
-import React from "react";
-import { Link } from "react-router-dom";
+import React from 'react';
+import { useNavigate } from 'react-router-dom';
+import "../styles/LoginPage.css";
 
-function LoginPage() {
+export default function Login() {
+  const [formData, setFormData] = React.useState({ email: "", password: "" })
+  const [item, setItems] = React.useState([])
+  const [formErrors, setFormErrors] = React.useState({});
+  const history = useNavigate();
+
+
+  React.useEffect(function () {
+    fetch('https://jsonplaceholder.typicode.com/users')
+      .then((response) => response.json())
+      .then(data => setItems(data))
+  }, [])
+
+
+  function handleChange(event) {
+    setFormData(prevFormData => {
+      return {
+        ...prevFormData,
+        [event.target.name]: event.target.value
+      };
+    });
+  }
+  const handleSubmit = (event) => {
+    event.preventDefault();
+    setFormErrors(validate(formData));
+  };
+
+  const validate = (values) => {
+    const errors = {};
+    const userid = formData.email;
+    const passid = formData.password;
+
+
+
+    item.map((items) => {
+      var paswd = items.username;
+      var num = 123;
+      paswd += num;
+      if (items.email !== userid && paswd !== passid) {
+        errors.username = "Invalid username or password";
+      }
+      else if (items.email === userid && paswd !== passid) {
+        errors.username = "Invalid username or passsword";
+      }
+      else if (items.email !== userid && paswd === passid) {
+        errors.username = "Invalid username or passsword";
+      }
+      else {
+        history("postDetails");
+      }
+    });
+    return errors;
+  };
```

Comment:
It's a good thing that you have abstracted the validation logic into a different function. You should be doing this more often.

2 things to note here:

1. Make sure your abstraction do only one thing.

> Here, validate function is validating the form + navigating to the next page. these two things should be kept separate. validate function should only validate and give you back the error details or a Boolean saying the form is valid or not.

2. try to move as many of the abstraction out of the react component.